drivers: display: add support for grayscale displays, st7528i driver

arbrauns opened the pull request:
This adds `PIXEL_FORMAT_GRAY4` and `PIXEL_FORMAT_GRAY16`, as well as a driver for the st7528i 16-level grayscale display controller.

arbrauns committed:
drivers: display: add st7528i driver

The ST7528 is a 160x100, 4-bit-grayscale dot matrix LCD controller.

According to the datasheet, the ST7528 supports SPI and parallel
interfaces, while the ST7528I supports only I2C.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns committed:
drivers: display: add gray-scale pixel formats

The draw buffers are represented as 4 (4-level grayscale)/2 (16-level
grayscale) horizontal pixels per byte, leftmost pixel starting at the most
significant bit.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+		ST7528_CMD_SET_COL_LSN | (start_col & 0x0f),
+	};
+
+	ret = st7528i_write_cmds(dev, cmd_buf, ARRAY_SIZE(cmd_buf));
```
What is the reason to use `ARRAY_SIZE()` on `uint8_t cmd_buf[]`?
`ret = st7528i_write_cmds(dev, cmd_buf, sizeof(cmd_buf));`

arbrauns replied:
Why not? It's the same thing, except that it also ensures that `cmd_buf` didn't accidentally decay to a pointer.

jfischer-no replied:
It is not the same, and also not from the code readability.

arbrauns commented on drivers/display/display_st7528.h:
```diff
@@ -0,0 +1,81 @@
+#define ST7528_CMD_DISPLAY_OFF 0xAE
+#define ST7528_CMD_DISPLAY_ON  0xAF
+
+#define ST7528_CMD_REVERSE_VIDEO_OFF 0xA6
+#define ST7528_CMD_REVERSE_VIDEO_ON  0xA7
```
[Reverse Video](https://en.wikipedia.org/wiki/Reverse_video) is a technical term referring to the inversion of background and foreground color, "reverse display" would be too easy to confuse with the options for horizontal and vertical flipping.

jfischer-no replied:
The description of this command is
"Reverse Display ON / OFF
Reverses the display status on LCD panel without rewriting the contents of the display data RAM."

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+struct st7528i_config {
+	struct i2c_dt_spec bus;
+	struct gpio_dt_spec reset;
+	struct gpio_dt_spec chip_select;
```
Why? That seems less clear to me with no benefit.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+/*
```
st7528 is the base model of the controller (SPI + parallel bus), st7528i is the I2C version. Register names and addresses are the same between the base model and I2C model, which is why I put them in `display_st7528.h`.

jfischer-no replied:
Usually the support for different supported interfaces is implemented in the same file, see ssd1306.c. So please keep it consistent.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+};
+
+struct st7528i_data {
+	uint8_t framebuffer[ST7528_NUM_PAGES * ST7528_BPP * ST7528_WIDTH];
```
Because setting the value of any one pixel also affects 7 neighbouring pixels. So either the driver has to keep track of the framebuffer itself, or perform a read of the controller framebuffer for every pixel that is set - since I2C reads seem to be broken (the datasheet kind of suggests that they aren't supported at all), the latter isn't even possible.

jfischer-no replied:
>Because setting the value of any one pixel also affects 7 neighbouring pixels.

That is not necessarily a reason if the tile areas are arbitrarily addressable.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+	int ret;
+
+	if (config->height != ST7528_HEIGHT) {
+		LOG_ERR("Invalid display height: expected %d, got %d", ST7528_HEIGHT,
+			config->height);
```
> if it is possible to wrap the line at an argument [...] so that the preferred line length is respected

What do you mean by that? The line you commented on *is* wrapped at an argument and the line length is respected. Can you elaborate on the style you would prefer?

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+	__ASSERT(col < ST7528_WIDTH, "width out of bounds");
+	__ASSERT(row < ST7528_HEIGHT, "height out of bounds");
+
+	const int page = row / 8;
+	const int page_mask = 1 << (row % 8);
+	uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+
+	uint8_t level = 0;
```
I've removed the empty line, but I can't place the `_ASSERT` below the variable definition without making it useless as a bounds check.

jfischer-no replied:
You can, by not using function parameters before.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+		k_sleep(ST7528_RESET_DELAY);
+	}
+
+	{
+		uint8_t cmd_buf[] = {
+			ST7528_CMD_SET_PARTIAL_DISPLAY_LINES,
+			config->com_partial_display,
+			ST7528_CMD_ADC_SELECT | config->invert_segments,
+			ST7528_CMD_SHL_SELECT | (config->invert_com << 3),
+			ST7528_CMD_SET_INITIAL_COM0,
+			config->com_offset,
+
+			ST7528_CMD_START_OSCILLATOR,
+			ST7528_CMD_SET_REGULATOR_RESISTOR | config->regulator_resistor,
+			ST7528_CMD_SET_ELECTRONIC_VOLUME,
+			config->electronic_volume,
+			ST7528_CMD_SET_LCD_BIAS | (config->lcd_bias - 5),
+
+			ST7528_CMD_SET_BOOST | 0x00,
+		};
+
+		ret = st7528i_write_cmds(dev, cmd_buf, ARRAY_SIZE(cmd_buf));
+		if (ret) {
+			return ret;
+		}
+	}
```
There aren't really any useful names to give here, this is simply the reset sequence as written in the datasheet - the only reason it's split up is because of the delays in the middle. The block scopes are necessary purely because of C's lack of shadowing and inline array/slice literals.

jfischer-no replied:
Well, please come up with concise functional names.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+		return ret;
+	}
+
+	for (int index = 0; index < 16; index++) {
```
The number of gray levels, I've added a `#define`.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+		 */
+		uint8_t cmd_buf[] = {
+			cmd | 0x00, value, cmd | 0x01, value, cmd | 0x02, value, cmd | 0x03, value,
+		};
```
I prefer a uniform style enforced by tooling - it's unfortunate that clang-format is too primitive to make this possible. I've added some clang-format exceptions to the arrays though.

jfischer-no replied:
NAK to `/* clang-format off */` nonsense, there is no enforcement to use clang-format.

arbrauns commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,513 @@
+		return -ENODEV;
+	}
+
+	if (config->width != ST7528_WIDTH) {
```
The bindings already ensure at compile time that this shouldn't happen (they have a `const` parameter). The controller does theoretically support an alternate display resolution, but that's not implemented here - this check is just an extra safeguard, I guess it could be removed entirely.

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,500 @@
+		return ret;
+	}
+
+	for (int index = 0; index < ST7528_NUM_GRAY_LEVELS; index++) {
+		uint8_t value = ST7528_PWM_LEVELS - (index * ST7528_PWM_LEVELS / 15);
+		uint8_t cmd = ST7528_CMD_SET_PALETTE_ENTRY | (index << 2);
```
```diff
diff --git a/drivers/display/display_st7528i.c b/drivers/display/display_st7528i.c
index 05f2c1dcba6..4fea545a58d 100644
--- a/drivers/display/display_st7528i.c
+++ b/drivers/display/display_st7528i.c
@@ -86,9 +86,9 @@ static int st7528i_write_palette(const struct device *dev)
                return ret;
        }
 
-       for (int index = 0; index < ST7528_NUM_GRAY_LEVELS; index++) {
-               uint8_t value = ST7528_PWM_LEVELS - (index * ST7528_PWM_LEVELS / 15);
-               uint8_t cmd = ST7528_CMD_SET_PALETTE_ENTRY | (index << 2);
+       for (int idx = 0; idx < ST7528_NUM_GRAY_LEVELS; idx++) {
+               uint8_t value = ST7528_PWM_LEVELS - (idx * ST7528_PWM_LEVELS / 15);
+               uint8_t cmd = ST7528_CMD_SET_PALETTE_ENTRY | (idx << 2);
 
                uint8_t cmd_buf[] = {
                        /*
```

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,500 @@
+	}
+
+	if (config->width != ST7528_WIDTH) {
+		LOG_ERR("Invalid display width: expected %d, got %d", ST7528_WIDTH, config->width);
```
```diff
diff --git a/drivers/display/display_st7528i.c b/drivers/display/display_st7528i.c
index 05f2c1dcba6..15cc8b8ccd0 100644
--- a/drivers/display/display_st7528i.c
+++ b/drivers/display/display_st7528i.c
@@ -198,15 +198,14 @@ static int st7528i_init(const struct device *dev)
        int ret;
 
        if (config->height != ST7528_HEIGHT) {
-               /* clang-format off */
                LOG_ERR("Invalid display height: expected %d, got %d",
                        ST7528_HEIGHT, config->height);
-               /* clang-format on */
                return -ENODEV;
        }
 
        if (config->width != ST7528_WIDTH) {
-               LOG_ERR("Invalid display width: expected %d, got %d", ST7528_WIDTH, config->width);
+               LOG_ERR("Invalid display width: expected %d, got %d",
+                       ST7528_WIDTH, config->width);
                return -ENODEV;
        }
```

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,500 @@
+}
+
+static inline void st7528i_set_pixel(struct st7528i_data *data, const uint16_t col,
+				     const uint16_t row, const uint8_t level)
+{
+	__ASSERT(col < ST7528_WIDTH, "width out of bounds");
+	__ASSERT(row < ST7528_HEIGHT, "height out of bounds");
+
+	const int page = row / 8;
+	const int page_mask = 1 << (row % 8);
+	uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+
+	for (int level_bit = 0; level_bit < ST7528_BPP; level_bit++) {
+		if (level & (1 << level_bit)) {
+			ptr[level_bit] |= page_mask;
+		} else {
+			ptr[level_bit] &= ~page_mask;
+		}
+	}
+}
+
+static inline uint8_t st7528i_get_pixel(struct st7528i_data *data, const uint16_t col,
+					const uint16_t row)
+{
+	__ASSERT(col < ST7528_WIDTH, "width out of bounds");
+	__ASSERT(row < ST7528_HEIGHT, "height out of bounds");
+
+	const int page = row / 8;
+	const int page_mask = 1 << (row % 8);
+	uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+	uint8_t level = 0;
+
+	for (int level_bit = 0; level_bit < ST7528_BPP; level_bit++) {
+		if (ptr[level_bit] & page_mask) {
+			level |= (1 << level_bit);
+		}
+	}
+
+	return level;
+}
```
```diff
diff --git a/drivers/display/display_st7528i.c b/drivers/display/display_st7528i.c
index 05f2c1dcba6..251195bd568 100644
--- a/drivers/display/display_st7528i.c
+++ b/drivers/display/display_st7528i.c
@@ -250,15 +250,17 @@ static int st7528i_init(const struct device *dev)
        return 0;
 }
 
-static inline void st7528i_set_pixel(struct st7528i_data *data, const uint16_t col,
-                                    const uint16_t row, const uint8_t level)
+static inline void st7528i_set_pixel(struct st7528i_data *data,
+                                    const uint16_t col,
+                                    const uint16_t row,
+                                    const uint8_t level)
 {
        __ASSERT(col < ST7528_WIDTH, "width out of bounds");
        __ASSERT(row < ST7528_HEIGHT, "height out of bounds");
 
        const int page = row / 8;
        const int page_mask = 1 << (row % 8);
-       uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+       uint8_t *ptr = &data->fb[(page * ST7528_WIDTH + col) * ST7528_BPP];
 
        for (int level_bit = 0; level_bit < ST7528_BPP; level_bit++) {
                if (level & (1 << level_bit)) {
@@ -269,7 +271,8 @@ static inline void st7528i_set_pixel(struct st7528i_data *data, const uint16_t c
        }
 }
 
-static inline uint8_t st7528i_get_pixel(struct st7528i_data *data, const uint16_t col,
+static inline uint8_t st7528i_get_pixel(struct st7528i_data *data,
+                                       const uint16_t col,
                                        const uint16_t row)
 {
        __ASSERT(col < ST7528_WIDTH, "width out of bounds");
@@ -277,7 +280,7 @@ static inline uint8_t st7528i_get_pixel(struct st7528i_data *data, const uint16_
 
        const int page = row / 8;
        const int page_mask = 1 << (row % 8);
-       uint8_t *ptr = &data->framebuffer[(page * ST7528_WIDTH + col) * ST7528_BPP];
+       uint8_t *ptr = &data->fb[(page * ST7528_WIDTH + col) * ST7528_BPP];
        uint8_t level = 0;
 
        for (int level_bit = 0; level_bit < ST7528_BPP; level_bit++) {
```

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,500 @@
+	return level;
+}
+
+static int st7528i_write_columns(const struct device *dev, const uint16_t page,
+				 const uint16_t start_col, const uint16_t num_cols)
+{
+	__ASSERT(page < ST7528_NUM_PAGES, "page out of bounds");
+	__ASSERT(start_col + num_cols <= ST7528_WIDTH, "columns out of bounds");
+
+	struct st7528i_data *data = dev->data;
+	int ret;
+
+	uint8_t cmd_buf[] = {
+		ST7528_CMD_SET_PAGE | page,
+		ST7528_CMD_SET_COL_MSN | (start_col >> 4),
+		ST7528_CMD_SET_COL_LSN | (start_col & 0x0f),
+	};
+
+	ret = st7528i_write_cmds(dev, cmd_buf, ARRAY_SIZE(cmd_buf));
+	if (ret) {
+		return ret;
+	}
+
+	return st7528i_write_data(
+		dev, &data->framebuffer[(page * ST7528_WIDTH + start_col) * ST7528_BPP],
+		num_cols * ST7528_BPP);
+}
```
```diff
diff --git a/drivers/display/display_st7528i.c b/drivers/display/display_st7528i.c
index 05f2c1dcba6..97dd79c16be 100644
--- a/drivers/display/display_st7528i.c
+++ b/drivers/display/display_st7528i.c
@@ -289,15 +289,17 @@ static inline uint8_t st7528i_get_pixel(struct st7528i_data *data, const uint16_
        return level;
 }
 
-static int st7528i_write_columns(const struct device *dev, const uint16_t page,
-                                const uint16_t start_col, const uint16_t num_cols)
+static int st7528i_write_columns(const struct device *dev,
+                                const uint16_t page,
+                                const uint16_t start_col,
+                                const uint16_t num_cols)
 {
-       __ASSERT(page < ST7528_NUM_PAGES, "page out of bounds");
-       __ASSERT(start_col + num_cols <= ST7528_WIDTH, "columns out of bounds");
-
        struct st7528i_data *data = dev->data;
        int ret;
 
+       __ASSERT(page < ST7528_NUM_PAGES, "page out of bounds");
+       __ASSERT(start_col + num_cols <= ST7528_WIDTH, "columns out of bounds");
+
        uint8_t cmd_buf[] = {
                ST7528_CMD_SET_PAGE | page,
                ST7528_CMD_SET_COL_MSN | (start_col >> 4),
@@ -310,7 +312,7 @@ static int st7528i_write_columns(const struct device *dev, const uint16_t page,
        }
 
        return st7528i_write_data(
-               dev, &data->framebuffer[(page * ST7528_WIDTH + start_col) * ST7528_BPP],
+               dev, &data->fb[(page * ST7528_WIDTH + start_col) * ST7528_BPP],
                num_cols * ST7528_BPP);
 }
```

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,500 @@
+	int ret;
+
+	const uint8_t pixels_per_byte = 8 / ST7528_BPP;
+
+	__ASSERT(desc->width <= desc->pitch, "Pitch is smaller than width");
+	__ASSERT((desc->pitch * desc->height) <= desc->buf_size, "Input buffer too small");
+	__ASSERT(desc->pitch % pixels_per_byte == 0, "Input buffer rows are not byte-aligned");
+
+	LOG_DBG("Writing %dx%d (w,h) @ %dx%d (x,y)", desc->width, desc->height, x_origin, y_origin);
+
+	const uint8_t *row_ptr = buf;
+	const int last_row = y_origin + desc->height - 1;
+
+	for (int row = y_origin; row <= last_row; row++) {
+		for (int x = 0; x < desc->width; x++) {
+			uint16_t col = x_origin + x;
+			uint8_t value;
+
+			if (x % pixels_per_byte == 0) {
+				value = (row_ptr[x / pixels_per_byte] & 0xf0) >> ST7528_BPP;
+			} else {
+				value = row_ptr[x / pixels_per_byte] & 0x0f;
+			}
+
+			st7528i_set_pixel(data, col, row, value);
+		}
+
+		if (row % 8 == 7 || row == last_row) {
+			/* last row of the current bank, send framebuffer to display */
+			ret = st7528i_write_columns(dev, row / 8, x_origin, desc->width);
+			if (ret) {
+				return ret;
+			}
+		}
+
+		row_ptr += (desc->pitch / pixels_per_byte);
+	}
+
+	return 0;
+}
```
```diff
diff --git a/drivers/display/display_st7528i.c b/drivers/display/display_st7528i.c
index 05f2c1dcba6..1b98ab9d026 100644
--- a/drivers/display/display_st7528i.c
+++ b/drivers/display/display_st7528i.c
@@ -332,40 +332,50 @@ static int st7528i_blanking_off(const struct device *dev)
        return st7528i_write_cmds(dev, cmd_buf, ARRAY_SIZE(cmd_buf));
 }
 
-static int st7528i_write(const struct device *dev, const uint16_t x_origin, const uint16_t y_origin,
-                        const struct display_buffer_descriptor *desc, const void *buf)
+static int st7528i_write(const struct device *dev,
+                        const uint16_t x_origin, const uint16_t y_origin,
+                        const struct display_buffer_descriptor *desc,
+                        const void *buf)
 {
+       const uint8_t pixels_per_byte = 8 / ST7528_BPP;
        struct st7528i_data *data = dev->data;
+       const uint8_t *row_ptr = buf;
+       const int last_row;
        int ret;
 
-       const uint8_t pixels_per_byte = 8 / ST7528_BPP;
-
-       __ASSERT(desc->width <= desc->pitch, "Pitch is smaller than width");
-       __ASSERT((desc->pitch * desc->height) <= desc->buf_size, "Input buffer too small");
-       __ASSERT(desc->pitch % pixels_per_byte == 0, "Input buffer rows are not byte-aligned");
+       __ASSERT(desc->width <= desc->pitch,
+                "Pitch is smaller than width");
+       __ASSERT((desc->pitch * desc->height) <= desc->buf_size,
+                "Input buffer too small");
+       __ASSERT(desc->pitch % pixels_per_byte == 0,
+                "Input buffer rows are not byte-aligned");
 
-       LOG_DBG("Writing %dx%d (w,h) @ %dx%d (x,y)", desc->width, desc->height, x_origin, y_origin);
+       last_row = y_origin + desc->height - 1;
 
-       const uint8_t *row_ptr = buf;
-       const int last_row = y_origin + desc->height - 1;
+       LOG_DBG("Writing %dx%d (w,h) @ %dx%d (x,y)",
+               desc->width, desc->height, x_origin, y_origin);
 
        for (int row = y_origin; row <= last_row; row++) {
                for (int x = 0; x < desc->width; x++) {
+                       uint8_t value = row_ptr[x / pixels_per_byte];
                        uint16_t col = x_origin + x;
-                       uint8_t value;
 
                        if (x % pixels_per_byte == 0) {
-                               value = (row_ptr[x / pixels_per_byte] & 0xf0) >> ST7528_BPP;
+                               value = (value & 0xf0) >> ST7528_BPP;
                        } else {
-                               value = row_ptr[x / pixels_per_byte] & 0x0f;
+                               value = value & 0x0f;
                        }
 
                        st7528i_set_pixel(data, col, row, value);
                }
 
                if (row % 8 == 7 || row == last_row) {
-                       /* last row of the current bank, send framebuffer to display */
-                       ret = st7528i_write_columns(dev, row / 8, x_origin, desc->width);
+                       /*
+                        * Last row of the current bank,
+                        * send framebuffer to display.
+                        */
+                       ret = st7528i_write_columns(dev, row / 8, x_origin,
+                                                   desc->width);
                        if (ret) {
                                return ret;
                        }

```

Apply similar for st7528i_read().

jfischer-no commented on drivers/display/display_st7528i.c:
```diff
@@ -0,0 +1,500 @@
+	.set_orientation = st7528i_set_orientation,
+};
+
+#define ST7528I_INIT(inst)                                                                         \
+	const static struct st7528i_config st7528i_config_##inst = {                               \
+		.bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.reset = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, {}),                          \
+		.chip_select = GPIO_DT_SPEC_INST_GET_OR(inst, chip_select_gpios, {}),              \
+		.height = DT_INST_PROP(inst, height),                                              \
+		.width = DT_INST_PROP(inst, width),                                                \
+		.framerate = DT_INST_PROP(inst, framerate),                                        \
+		.com_partial_display = DT_INST_PROP(inst, com_partial_display),                    \
+		.com_offset = DT_INST_PROP(inst, com_offset),                                      \
+		.invert_com = DT_INST_PROP(inst, invert_com),                                      \
+		.invert_segments = DT_INST_PROP(inst, invert_segments),                            \
+		.lcd_bias = DT_INST_PROP(inst, lcd_bias),                                          \
+		.regulator_resistor = DT_INST_PROP(inst, regulator_resistor),                      \
+		.electronic_volume = DT_INST_PROP(inst, electronic_volume),                        \
+		.boost = DT_INST_PROP(inst, boost),                                                \
+	};                                                                                         \
+                                                                                                   \
+	static struct st7528i_data st7528i_data_##inst = {                                         \
+		.framebuffer = {0},                                                                \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, st7528i_init, NULL, &st7528i_data_##inst,                      \
+			      &st7528i_config_##inst, POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY,   \
+			      &st7528i_api);
+
```
```c
diff --git a/drivers/display/display_st7528i.c b/drivers/display/display_st7528i.c
index 05f2c1dcba6..2fb42423698 100644
--- a/drivers/display/display_st7528i.c
+++ b/drivers/display/display_st7528i.c
@@ -478,30 +478,31 @@ static const struct display_driver_api st7528i_api = {
        .set_orientation = st7528i_set_orientation,
 };
 
-#define ST7528I_INIT(inst)                                                                         \
-       const static struct st7528i_config st7528i_config_##inst = {                               \
-               .bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
-               .reset = GPIO_DT_SPEC_INST_GET_OR(inst, reset_gpios, {}),                          \
-               .chip_select = GPIO_DT_SPEC_INST_GET_OR(inst, chip_select_gpios, {}),              \
-               .height = DT_INST_PROP(inst, height),                                              \
-               .width = DT_INST_PROP(inst, width),                                                \
-               .framerate = DT_INST_PROP(inst, framerate),                                        \
-               .com_partial_display = DT_INST_PROP(inst, com_partial_display),                    \
-               .com_offset = DT_INST_PROP(inst, com_offset),                                      \
-               .invert_com = DT_INST_PROP(inst, invert_com),                                      \
-               .invert_segments = DT_INST_PROP(inst, invert_segments),                            \
-               .lcd_bias = DT_INST_PROP(inst, lcd_bias),                                          \
-               .regulator_resistor = DT_INST_PROP(inst, regulator_resistor),                      \
-               .electronic_volume = DT_INST_PROP(inst, electronic_volume),                        \
-               .boost = DT_INST_PROP(inst, boost),                                                \
-       };                                                                                         \
-                                                                                                   \
-       static struct st7528i_data st7528i_data_##inst = {                                         \
-               .framebuffer = {0},                                                                \
-       };                                                                                         \
-                                                                                                   \
-       DEVICE_DT_INST_DEFINE(inst, st7528i_init, NULL, &st7528i_data_##inst,                      \
-                             &st7528i_config_##inst, POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY,   \
+#define ST7528I_INIT(n)                                                                \
+       const static struct st7528i_config st7528i_config_##n = {               \
+               .bus = I2C_DT_SPEC_INST_GET(n),                                 \
+               .reset = GPIO_DT_SPEC_INST_GET_OR(n, reset_gpios, {}),          \
+               .cs = GPIO_DT_SPEC_INST_GET_OR(n, cs_gpios, {}),                \
+               .height = DT_INST_PROP(n, height),                              \
+               .width = DT_INST_PROP(n, width),                                \
+               .framerate = DT_INST_PROP(n, framerate),                        \
+               .com_partial_display = DT_INST_PROP(n, com_partial_display),    \
+               .com_offset = DT_INST_PROP(n, com_offset),                      \
+               .invert_com = DT_INST_PROP(n, invert_com),                      \
+               .invert_segments = DT_INST_PROP(n, invert_segments),            \
+               .lcd_bias = DT_INST_PROP(n, lcd_bias),                          \
+               .regulator_resistor = DT_INST_PROP(n, regulator_resistor),      \
+               .electronic_volume = DT_INST_PROP(n, electronic_volume),        \
+               .boost = DT_INST_PROP(n, boost),                                \
+       };                                                                      \
+                                                                               \
+       static struct st7528i_data st7528i_data_##n = {                         \
+               .framebuffer = {0},                                             \
+       };                                                                      \
+                                                                               \
+       DEVICE_DT_INST_DEFINE(n, st7528i_init, NULL,                            \
+                             &st7528i_data_##n,&st7528i_config_##n,            \
+                             POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY,        \
                              &st7528i_api);
 
 DT_INST_FOREACH_STATUS_OKAY(ST7528I_INIT)
```

jfischer-no commented on dts/bindings/display/sitronix,st7528i.yaml:
```diff
@@ -0,0 +1,114 @@
+
+      If connected directly the MCU pin should be configured as active low.
+
+  chip-select-gpios:
```
Datasheet name for this pin is CSB, name should be csb-gpios or just use common abbreviation for chip select cs-gpios.

arbrauns commented:
I don't have the time to comply with nitpicks like a bunch of manual formatting, please feel free to land this yourself.

arbrauns committed:
lvgl: add support for 2/4-bit gray-scale pixel formats

They are represented in the draw buffer as 4/2 horizontal pixels per byte,
leftmost pixel at MSB.

Signed-off-by: Armin Brauns <armin.brauns@embedded-solutions.at>

arbrauns commented:
Rebased on `main`, and I think most of the substantial review points have been addressed - if you want to enforce your personal style preferences, you have push access to my branch, so please feel free to do so directly.

faxe1008 commented on modules/lvgl/lvgl_display_gray.c:
```diff
@@ -0,0 +1,109 @@
+	switch (data->cap.current_pixel_format) {
+	case PIXEL_FORMAT_GRAY4: {
+#ifdef CONFIG_LV_Z_BUFFER_ALLOC_STATIC
+		__ASSERT(CONFIG_LV_Z_BITS_PER_PIXEL >= 2, "draw buffer too small");
```
Can you make those BUILD_CHECKS instead? This can be caught during build time.

arbrauns replied:
Not really, since they depend on the `current_pixel_format`.

faxe1008 replied:
This is possible:
```
#ifdef CONFIG_LV_Z_BUFFER_ALLOC_STATIC
		BUILD_ASSERT(CONFIG_LV_Z_BITS_PER_PIXEL >= 2, "draw buffer too small");
#endif
```

arbrauns replied:
No, because then LVGL would fail to build even if the application never uses `PIXEL_FORMAT_GRAY4`. `CONFIG_LV_Z_BITS_PER_PIXEL == 1` is fine as long as you only ever use `PIXEL_FORMAT_MONO01`.

faxe1008 replied:
Ah, yes ok that's what you meant. You are right.

faxe1008 commented on modules/lvgl/lvgl_display_gray.c:
```diff
@@ -0,0 +1,109 @@
+	}
+	case PIXEL_FORMAT_GRAY16: {
+#ifdef CONFIG_LV_Z_BUFFER_ALLOC_STATIC
+		__ASSERT(CONFIG_LV_Z_BITS_PER_PIXEL >= 4, "draw buffer too small");
```
Here also.

faxe1008 commented on modules/lvgl/lvgl_display_gray.c:
```diff
@@ -0,0 +1,109 @@
+	}
+	default:
+		__ASSERT(false, "Unsupported pixel format");
+		return;
```
Return is not needed here.

arbrauns replied:
`__ASSERT()` is a no-op by default, which would make omitting the `return` UB.

faxe1008 replied:
Since the function does not have any return value, I don't see how this should be UB. If the `__ASSERT` is no-op this will fall through and reach the end of the function, without any side effects. 

ISO/IEC 9899:1999 section 6.8.4.2 does not show any restrictions to the statement within the `default` clause (only that there is only one at most).

arbrauns replied:
Oh, sorry, I was looking at the wrong thing - on [L31](https://github.com/zephyrproject-rtos/zephyr/pull/53596/files/b83ff886737a4da6f26188b1c42dd6c706c7e9e1#diff-a1f332288b9aa993f0c8f660df953dfdbe24864bb3ba548ba80d66778c00586cR31), removing the `return` would lead to `pixels_per_byte` being uninitialized. Personally I'd still keep it in the spirit of defensive programming, but I've removed it.

faxe1008 commented on modules/lvgl/lvgl_display_gray.c:
```diff
@@ -0,0 +1,109 @@
+	}
+	default:
+		__ASSERT(false, "Unsupported pixel format");
+		return;
```
Return is not needed

arbrauns commented:
Not stale, waiting on review.

danieldegrasse commented:
> Not stale, waiting on review.

@jfischer-no can you revisit? I assume the primary block is over the column width of some lines. Zephyr's requirement is 100 columns or less per line, but some reviewers will request 80 columns (this should be a *request*, not a block). @arbrauns, if you want to apply this style without manually fixing line lengths, you can always edit the Zephyr `.clang-format` file like so:

```diff
diff --git a/.clang-format b/.clang-format
index 15012747148..b134f27fff7 100644
--- a/.clang-format
+++ b/.clang-format
@@ -28,7 +28,7 @@ AttributeMacros:
   - __subsystem
 BitFieldColonSpacing: After
 BreakBeforeBraces: Linux
-ColumnLimit: 100
+ColumnLimit: 80
 ConstructorInitializerIndentWidth: 8
 ContinuationIndentWidth: 8
 ForEachMacros:
```

Then you can run `clang-format -i` on the display driver file to shorten all lines to 80 columns or less.

jfischer-no reviewed (changes requested):
Not addressing comments or just marking them resolved, and not following style how we write the driver does not help, my -1 stays. Also, the code is outdated, explicit return of -ENOTSUP is not required.
